### Part 1

In [625]:
from sklearn.preprocessing import LabelEncoder as le # For label encoding
import category_encoders as ce # For category encoding
import pandas as pd # To manipulate dataset
import numpy as np # To use different functions like replace

In [626]:
data = pd.read_csv('vehicles_2019.csv') # Import the csv file
data.transpose() # View it by interchanging rows and columns 

0           1     \
Unnamed: 0                                              0           1   
Model Year                                           2019        2019   
Mfr Name                                            Honda  FCA US LLC   
Division                                            Acura  ALFA ROMEO   
Carline                                               NSX          4C   
...                                                   ...         ...   
120V Charge time at 120 Volts (hours)                 NaN         NaN   
PHEV Total Driving Range (rounded to nearest 10...    NaN         NaN   
City PHEV Composite MPGe                              NaN         NaN   
Hwy PHEV Composite MPGe                               NaN         NaN   
Comb PHEV Composite MPGe                              NaN         NaN   

                                                                        2     \
Unnamed: 0                                                                 2   
Model Year                                                              2019   
Mfr Name                                                        aston martin   
Division                                            Aston Martin Lagonda Ltd   
Carline                                                           Vantage V8   
...                                                                      ...   
120V Charge time at 120 Volts (hours)                                    NaN   
PHEV Total Driving Range (rounded to nearest 10...                       NaN   
City PHEV Composite MPGe                                                 NaN   
Hwy PHEV Composite MPGe                                                  NaN   
Comb PHEV Composite MPGe                                                 NaN   

                                                                   3     \
Unnamed: 0                                                            3   
Model Year                                                         2019   
Mfr Name                                            Volkswagen Group of   
Division                                                           Audi   
Carline                                             TT Roadster quattro   
...                                                                 ...   
120V Charge time at 120 Volts (hours)                               NaN   
PHEV Total Driving Range (rounded to nearest 10...                  NaN   
City PHEV Composite MPGe                                            NaN   
Hwy PHEV Composite MPGe                                             NaN   
Comb PHEV Composite MPGe                                            NaN   

                                                            4     \
Unnamed: 0                                                     4   
Model Year                                                  2019   
Mfr Name                                                     BMW   
Division                                                     BMW   
Carline                                             Z4 sDrive30i   
...                                                          ...   
120V Charge time at 120 Volts (hours)                        NaN   
PHEV Total Driving Range (rounded to nearest 10...           NaN   
City PHEV Composite MPGe                                     NaN   
Hwy PHEV Composite MPGe                                      NaN   
Comb PHEV Composite MPGe                                     NaN   

                                                                   5     \
Unnamed: 0                                                            5   
Model Year                                                         2019   
Mfr Name                                            Volkswagen Group of   
Division                                                        Bugatti   
Carline                                                          Chiron   
...                                            

In [627]:
new_data=data[['Mfr Name','Division','Carline','Verify Mfr Cd','Index (Model Type Index)','Eng Displ','# Cyl','Transmission',
               'Comb Unadj FE - Conventional Fuel','Air Aspiration Method Desc','Trans Desc','# Gears',
               'Lockup Torque Converter','Trans Creeper Gear','Drive Sys','Max Ethanol % - Gasoline','Max Biodiesel %',
                'Fuel Usage Desc - Conventional Fuel',
                'Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)','2Dr Pass Vol','2Dr Lugg Vol','4Dr Pass Vol',
                '4Dr Lugg Vol','Htchbk Pass Vol','Htchbk Lugg Vol','Intake Valves Per Cyl','Exhaust Valves Per Cyl','Carline Class',
                'Carline Class Desc','Calc Approach Desc','Cyl Deact?','Var Valve Timing?','Var Valve Timing Desc','Var Valve Lift?',
                 'Fuel Metering Sys Desc','Fuel Metering Sys Cd','Oil Viscosity','Stop/Start System (Engine Management System) Code',
                  'FE Rating (1-10 rating on Label)','GHG Rating (1-10 rating on Label)',
                 'GHG 1-10 rating on Ethanol (EPA Determined)','#1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 1)',
                 'City CO2 Rounded Adjusted','Hwy CO2 Rounded Adjusted','Comb CO2 Rounded Adjusted (as shown on FE Label)']]

new_data.transpose() # Looking at the data by interchanging rows and columns 

0     \
Mfr Name                                                                                        Honda   
Division                                                                                        Acura   
Carline                                                                                           NSX   
Verify Mfr Cd                                                                                     HNX   
Index (Model Type Index)                                                                           41   
Eng Displ                                                                                         3.5   
# Cyl                                                                                               6   
Transmission                                                                              Auto(AM-S9)   
Comb Unadj FE - Conventional Fuel                                                             29.3561   
Air Aspiration Method Desc                                                               Turbocharged   
Trans Desc                                          Automated Manual- Selectable (e.g. Automated M...   
# Gears                                                                                             9   
Lockup Torque Converter                                                                             Y   
Trans Creeper Gear                                                                                  N   
Drive Sys                                                                                           A   
Max Ethanol % - Gasoline                                                                           10   
Max Biodiesel %                                                                                   NaN   
Fuel Usage Desc - Conventional Fuel                              Gasoline (Premium Unleaded Required)   
Gas Guzzler Exempt (Where Truck = 1975 NHTSA tr...                                                  N   
2Dr Pass Vol                                                                                      NaN   
2Dr Lugg Vol                                                                                      NaN   
4Dr Pass Vol                                                                                      NaN   
4Dr Lugg Vol                                                                                      NaN   
Htchbk Pass Vol                                                                                   NaN   
Htchbk Lugg Vol                                                                                   NaN   
Intake Valves Per Cyl                                                                               2   
Exhaust Valves Per Cyl                                                                              2   
Carline Class                                                                                       1   
Carline Class Desc                                                                        Two Seaters   
Calc Approach Desc                                                              Derived 5-cycle label   
Cyl Deact?                                                                                          N   
Var Valve Timing?                                                                                   Y   
Var Valve Timing Desc                                Variable Timing control (VTC) system is applied.   
Var Valve Lift?                                                                                     N   
Fuel Metering Sys Desc                                                Spark Ignition Direct Injection   
Fuel Metering Sys Cd                                                                              GDI   
Oil Viscosity                                                                                   0W-40   
Stop/Start System (Engine Management System) Code                                                   Y   
FE

In [628]:
new_data.isnull().sum() # checking number of null values

Mfr Name                                                                  0
Division                                                                  0
Carline                                                                   0
Verify Mfr Cd                                                             0
Index (Model Type Index)                                                  0
Eng Displ                                                                 0
# Cyl                                                                     0
Transmission                                                              0
Comb Unadj FE - Conventional Fuel                                         0
Air Aspiration Method Desc                                                0
Trans Desc                                                                0
# Gears                                                                   0
Lockup Torque Converter                                                   0
Trans Creepe

In [629]:
new_data.describe() # looking at the features in terms of mean, max, and etc

Index (Model Type Index)    Eng Displ       # Cyl  \
count               1270.000000  1270.000000  1270.00000   
mean                 224.362205     3.066772     5.54252   
std                  226.132519     1.281498     1.76455   
min                    1.000000     1.000000     3.00000   
25%                   41.000000     2.000000     4.00000   
50%                  127.000000     3.000000     6.00000   
75%                  400.000000     3.600000     6.00000   
max                  991.000000     8.000000    16.00000   

       Comb Unadj FE - Conventional Fuel      # Gears  \
count                        1270.000000  1270.000000   
mean                           30.991457     7.109449   
std                             8.652050     1.754379   
min                            12.885000     1.000000   
25%                            24.975525     6.000000   
50%                            29.755150     7.000000   
75%                            34.962775     8.000000   
max                            80.847200    10.000000   

       Max Ethanol % - Gasoline  Max Biodiesel %  2Dr Pass Vol  2Dr Lugg Vol  \
count               1240.000000        30.000000    183.000000    183.000000   
mean                  14.391129        17.500000     81.557377      9.131148   
std                   13.597028         5.685735      8.980794      3.302390   
min                   10.000000         5.000000     43.000000      0.000000   
25%                   10.000000        20.000000     76.000000      5.000000   
50%                   10.000000        20.000000     82.000000      9.000000   
75%                   15.000000        20.000000     89.000000     11.000000   
max                   85.000000        20.000000    100.000000     22.000000   

       4Dr Pass Vol  ...  Intake Valves Per Cyl  Exhaust Valves Per Cyl  \
count    417.000000  ...            1270.000000             1270.000000   
mean      99.311751  ...               1.923622                1.920472   
std        7.208681  ...               0.265707                0.270667   
min       75.000000  ...               1.000000                1.000000   
25%       94.000000  ...               2.000000                2.000000   
50%       99.000000  ...               2.000000                2.000000   
75%      104.000000  ...               2.000000                2.000000   
max      127.000000  ...               2.000000                2.000000   

       Carline Class  FE Rating (1-10 rating on Label)  \
count    1270.000000                       1270.000000   
mean       12.937795                          4.576378   
std        12.017720                          1.668815   
min         1.000000                          1.000000   
25%         4.000000                          3.000000   
50%         6.000000                          4.000000   
75%        30.000000                          5.000000   
max        33.000000                         10.000000   

       GHG Rating (1-10 rating on Label)  \
count                        1270.000000   
mean                            4.554331   
std                             1.655935   
min                             1.000000   
25%                             3.000000   
50%                             4.000000   
75%                             5.000000   
max                            10.000000   

       GHG 1-10 rating on Ethanol (EPA Determined)  \
count                                    43.000000   
mean                                      3.604651   
std                                       1.003317   
min                                       1.000000   
25%                                       3.000000   
50%                                       3.000000   
75%                                       4.000000   
max                                       6.000000   

       #1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 1)  \
count                                        1270.000000              

In [630]:
new_data['Mfr Name'].value_counts(dropna=False) # Looking at the unique values 

General Motors          190
Ford Motor Company      122
BMW                     122
Toyota                  101
Mercedes-Benz            91
FCA US LLC               91
Porsche                  72
Volkswagen Group of      69
Nissan                   68
Jaguar Land Rover L      64
Honda                    64
Hyundai                  55
Kia                      34
MAZDA                    23
Subaru                   22
Volvo                    20
Mitsubishi Motors Co     16
Maserati                 11
Ferrari                  10
McLaren Automotive        7
Rolls-Royce               7
aston martin              6
Roush                     4
SUBARU TECNICA INTE       1
Name: Mfr Name, dtype: int64

In [631]:
new_data['Division'].value_counts(dropna=False) # Looking at the unique values 

Ford                                95
Chevrolet                           94
Mercedes-Benz                       91
BMW                                 87
Porsche                             72
TOYOTA                              62
GMC                                 52
Honda                               46
NISSAN                              44
Jaguar                              42
HYUNDAI MOTOR COMPANY               40
LEXUS                               39
Mini                                35
KIA MOTORS CORPORATION              34
Audi                                33
Jeep                                31
Cadillac                            27
Lincoln                             27
Volkswagen                          26
Dodge                               24
INFINITI                            24
MAZDA                               23
Subaru                              22
Land Rover                          22
Volvo Cars of North America, LLC    20
Acura                    

In [632]:
new_data['Carline'].value_counts(dropna=False) # Looking at the unique values 

CAMARO                      8
K10 SIERRA 4WD              7
C10 SILVERADO 2WD           7
C10 SIERRA 2WD              7
K10 SILVERADO 4WD           7
                           ..
Sonata HYBRID SE            1
GLE 400 4MATIC              1
AMG C 43 4MATIC             1
QX60 FWD                    1
340i xDrive Gran Turismo    1
Name: Carline, Length: 848, dtype: int64

In [633]:
new_data['Air Aspiration Method Desc'].value_counts() # Looking at the unique values 

Turbocharged                 679
Naturally Aspirated          524
Supercharged                  55
Turbocharged+Supercharged      7
Other                          5
Name: Air Aspiration Method Desc, dtype: int64

In [634]:
new_data['Transmission'].value_counts(dropna=False) # Looking at the unique values 

Auto(S8)        284
Auto(S6)        164
Manual(M6)      117
Auto(A9)        101
Auto(A8)         83
Auto(A6)         81
Auto(AM-S7)      70
Auto(S10)        59
Auto(AV)         50
Auto(AM7)        36
Auto(AV-S7)      30
Auto(S7)         26
Auto(S9)         25
Auto(AV-S6)      21
Auto(A7)         21
Manual(M7)       19
Auto(A10)        15
Auto(AM-S8)      15
Manual(M5)       14
Auto(AM6)        10
Auto(AV-S8)       9
Auto(AV-S10)      6
Auto(A5)          5
Auto(AM-S6)       3
Auto(S5)          3
Auto(A4)          2
Auto(AM-S9)       1
Name: Transmission, dtype: int64

In [635]:
new_data['Trans Desc'].value_counts(dropna=False) # Looking at the unique values 

Semi-Automatic                                                       561
Automatic                                                            308
Manual                                                               150
Automated Manual- Selectable (e.g. Automated Manual with paddles)     89
Selectable Continuously Variable (e.g. CVT with paddles)              66
Continuously Variable                                                 50
Automated Manual                                                      46
Name: Trans Desc, dtype: int64

In [636]:
new_data['Lockup Torque Converter'].value_counts(dropna=False) # Looking at the unique values 

Y    1014
N     256
Name: Lockup Torque Converter, dtype: int64

In [637]:
new_data['Drive Sys'].value_counts(dropna=False) # Looking at the unique values 

A    355
R    353
F    346
4    176
P     40
Name: Drive Sys, dtype: int64

In [638]:
new_data['Trans Creeper Gear'].value_counts(dropna=False) # Looking at the unique values 

N    1269
Y       1
Name: Trans Creeper Gear, dtype: int64

In [639]:
new_data['Max Ethanol % - Gasoline'].value_counts(dropna=False) # Looking at the unique values 

10.0    753
15.0    444
85.0     43
NaN      30
Name: Max Ethanol % - Gasoline, dtype: int64

In [640]:
new_data['Max Biodiesel %'].value_counts(dropna=False) # Looking at the unique values 

NaN     1240
20.0      25
5.0        5
Name: Max Biodiesel %, dtype: int64

In [641]:
new_data['Fuel Usage Desc - Conventional Fuel'].value_counts(dropna=False) # Looking at the unique values 

Gasoline (Regular Unleaded Recommended)       592
Gasoline (Premium Unleaded Required)          323
Gasoline (Premium Unleaded Recommended)       309
Diesel, ultra low sulfur (15 ppm, maximum)     30
Gasoline (Mid Grade Unleaded Recommended)      16
Name: Fuel Usage Desc - Conventional Fuel, dtype: int64

In [642]:
new_data['Fuel Usage Desc - Conventional Fuel'].value_counts(dropna=False) # Looking at the unique values 

Gasoline (Regular Unleaded Recommended)       592
Gasoline (Premium Unleaded Required)          323
Gasoline (Premium Unleaded Recommended)       309
Diesel, ultra low sulfur (15 ppm, maximum)     30
Gasoline (Mid Grade Unleaded Recommended)      16
Name: Fuel Usage Desc - Conventional Fuel, dtype: int64

In [643]:
new_data['Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)'].value_counts(dropna=False) # Looking at the unique values 

N    752
T    518
Name: Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition), dtype: int64

In [644]:
new_data['2Dr Pass Vol'].value_counts(dropna=False) # Looking at the unique values 

NaN      1087
70.0       24
90.0       18
68.0       13
91.0       12
89.0       12
77.0       12
81.0       10
79.0        9
82.0        9
94.0        8
85.0        7
84.0        7
83.0        7
76.0        6
99.0        5
86.0        4
88.0        4
80.0        3
87.0        3
74.0        3
72.0        3
100.0       1
75.0        1
78.0        1
43.0        1
Name: 2Dr Pass Vol, dtype: int64

In [645]:
new_data['2Dr Lugg Vol'].unique() # Looking at the unique values 

array([nan,  0.,  9.,  5.,  7., 10., 12., 13., 15.,  8., 14., 11., 22.,
       16.])

In [646]:
new_data['4Dr Pass Vol'].unique() # Looking at the unique values 

array([ nan,  83.,  86.,  85.,  91.,  89.,  93.,  94.,  92.,  90.,  97.,
        96.,  88., 102., 100.,  99., 101., 103.,  98., 106., 104., 105.,
        95., 108., 111., 114., 110., 113., 109., 112., 120., 117., 116.,
       127.,  75.])

In [647]:
new_data['4Dr Lugg Vol'].unique() # Looking at the unique values 

array([nan, 14., 12., 10., 11., 13.,  9., 20., 17., 15., 16., 23., 18.,
       19., 27., 28., 22., 24., 21., 30., 29., 34., 35., 42., 26., 25.,
       47.])

In [648]:
new_data['Htchbk Pass Vol'].unique() # Looking at the unique values 

array([ nan,  69.,  76.,  85.,  80.,  84.,  91.,  89.,  90.,  88.,  87.,
        86.,  94.,  93.,  95.,  98.,  92.,  97., 102., 107., 101.,  96.])

In [649]:
new_data['Htchbk Lugg Vol'].unique() # Looking at the unique values 

array([nan,  7., 15.,  9., 22., 19., 17., 16., 18., 25., 21., 20., 27.,
       31., 26.])

In [650]:
# Dont need any encoding
new_data['Intake Valves Per Cyl'].value_counts(dropna=False) # Looking at the unique values 

2    1173
1      97
Name: Intake Valves Per Cyl, dtype: int64

In [651]:
# Dont need any encoding
new_data['Exhaust Valves Per Cyl'].value_counts(dropna=False) # Looking at the unique values 

2    1169
1     101
Name: Exhaust Valves Per Cyl, dtype: int64

In [652]:
# Dont need any encoding
new_data['Intake Valves Per Cyl'].value_counts(dropna=False) # Looking at the unique values 

2    1173
1      97
Name: Intake Valves Per Cyl, dtype: int64

In [653]:
new_data['Carline Class Desc'].value_counts(dropna=False) # Looking at the unique values 

Midsize Cars                            189
Compact Cars                            141
Small SUV 4WD                           135
Subcompact Cars                         116
Large Cars                              115
Standard SUV 4WD                         99
Two Seaters                              85
Small SUV 2WD                            83
Minicompact Cars                         58
Standard Pick-up Trucks 4WD              51
Standard Pick-up Trucks 2WD              42
Standard SUV 2WD                         36
Small Station Wagons                     35
Small Pick-up Trucks 4WD                 17
Special Purpose Vehicle 2WD              16
Small Pick-up Trucks 2WD                 16
Midsize Station Wagons                   11
Special Purpose Vehicle cab chassis      10
Special Purpose Vehicle, minivan 2WD      8
Vans, Passenger Type                      3
Special Purpose Vehicle 4WD               3
Special Purpose Vehicle, minivan 4WD      1
Name: Carline Class Desc, dtype:

In [654]:
new_data['Calc Approach Desc'].value_counts(dropna=False) # Looking at the unique values 

Derived 5-cycle label                                                                                                               1043
Vehicle Specific 5-cycle label                                                                                                       132
Derived Vehicle Specific 5-cycle Calculation Approach for city label but Modified 5-cycle Calculation Approach for Highway label      95
Name: Calc Approach Desc, dtype: int64

In [655]:
new_data['Cyl Deact?'].value_counts(dropna=False) # Looking at the unique values 

N    1096
Y     174
Name: Cyl Deact?, dtype: int64

In [656]:
new_data['Var Valve Timing?'].value_counts(dropna=False) # Looking at the unique values 

Y    1230
N      40
Name: Var Valve Timing?, dtype: int64

In [657]:
new_data['Var Valve Timing Desc'].value_counts(dropna=False) # Looking at the unique values 

variable valve timing at inlet and outlet valves                                                                    104
Intake and exhaust                                                                                                   91
cam phaser                                                                                                           76
Porsche VVT                                                                                                          72
Allows timing of the intake / exhaust valves to be changed while the engine is in operation                          53
                                                                                                                   ... 
Variation of opening and closing of intake and exhaust valves on a single camshaft. No change in valve overlaps.      1
Variable Valve Timing (VTC) and Lift Electronic Control (VTEC) system is applied.                                     1
Variable Valve Timing and Lift Electroni

In [658]:
new_data['Var Valve Lift?'].value_counts(dropna=False) # Looking at the unique values 

N    898
Y    372
Name: Var Valve Lift?, dtype: int64

In [659]:
new_data['Fuel Metering Sys Cd'].value_counts(dropna=False) # Looking at the unique values 

GDI     897
MFI     237
GDPI    106
CRDI     30
Name: Fuel Metering Sys Cd, dtype: int64

In [660]:
new_data['Fuel Metering Sys Desc'].value_counts(dropna=False) # Looking at the unique values 

Spark Ignition Direct Injection             897
Multipoint/sequential fuel injection        237
Spark Ignition direct & ported injection    106
Common Rail Direct Diesel Injection          30
Name: Fuel Metering Sys Desc, dtype: int64

In [661]:
new_data['Oil Viscosity'].value_counts(dropna=False) # Looking at the unique values 

0W20             198
5W30             182
0W-20            139
0W40              98
5W20 GF4          49
                ... 
5W20 API SM        1
SAE 0W-40          1
SAE 5W-50          1
SAE 0W30           1
5W-30 API SN+      1
Name: Oil Viscosity, Length: 84, dtype: int64

In [662]:
new_data['Stop/Start System (Engine Management System) Code'].value_counts(dropna=False) # Looking at the unique values 

Y    659
N    611
Name: Stop/Start System (Engine Management System) Code, dtype: int64

In [663]:
new_data['GHG 1-10 rating on Ethanol (EPA Determined)'].unique() # Looking at the unique values 

array([nan,  6.,  5.,  4.,  3.,  1.,  2.])

In [664]:
# Frequency/count encoding for Mfr Name
encoder = ce.CountEncoder(cols=['Mfr Name'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Frequency/count encoding for Division
encoder = ce.CountEncoder(cols=['Division'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Ordinal encoding for air aspiration 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Air Aspiration Method Desc', 'mapping': {
                                    'Supercharged': 1, 'Turbocharged': 2, 
                                    'Other':3, 'Naturally Aspirated':4}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Frequency/count encoding for carline
encoder = ce.CountEncoder(cols=['Carline'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Ordinal encoding for Trans Desc 
new_data['Trans Desc'].unique()
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Trans Desc', 'mapping': {
                                    'Manual': 1, 'Automated Manual': 2, 
                                    'Semi-Automatic':3, 'Automated Manual- Selectable (e.g. Automated Manual with paddles)':4,
                                    'Automatic': 5,'Selectable Continuously Variable (e.g. CVT with paddles)': 6,
                                    'Continuously Variable':7 }}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Frequency/count encoding for Verify Mfr Cd
encoder = ce.CountEncoder(cols=['Verify Mfr Cd'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Lockup Torque Converter 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Lockup Torque Converter',
                                      'mapping': {'Y': 0, 'N': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Lockup Torque Converter 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Drive Sys', 
                                      'mapping': {'A': 0, 'R': 1, 'F': 2, '4': 3, 'P': 4}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Frequency/count encoding for Calc Approach Desc
encoder = ce.CountEncoder(cols=['Calc Approach Desc'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Trans Creeper Gear 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Trans Creeper Gear', 
                                      'mapping': {'Y': 0, 'N': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Fuel Usage Desc - Conventional Fuel
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Fuel Usage Desc - Conventional Fuel', 
                            'mapping':{'Gasoline (Regular Unleaded Recommended)': 0,
                                    'Gasoline (Premium Unleaded Required)': 1,
                                    'Gasoline (Premium Unleaded Recommended)': 2,
                                    'Diesel, ultra low sulfur (15 ppm, maximum)': 3,
                                    'Gasoline (Mid Grade Unleaded Recommended)': 4}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)
new_data['Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)'].unique()
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)', 
                                      'mapping': {'T': 0, 'N': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Frequency/count encoding for Carline Class Desc
encoder = ce.CountEncoder(cols=['Carline Class Desc'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Cyl Deact? 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Cyl Deact?', 'mapping': {'Y': 0, 'N': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Ordinal encoding for Var Valve Timing? 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Var Valve Timing?', 'mapping': {'N': 0, 'Y': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Var Var Valve Timing Desc
new_data['Var Valve Timing Desc'] = new_data['Var Valve Timing Desc'].astype('category') # converting type of columns to 'category'
new_data['Var Valve Timing Desc'] = new_data['Var Valve Timing Desc'].cat.codes # Assigning numerical values and storing in another column

# Ordinal encoding for Var Valve Timing? 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Var Valve Lift?', 'mapping': {'N': 0, 'Y': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Fuel Metering Sys Cd
new_data['Fuel Metering Sys Cd'] = new_data['Fuel Metering Sys Cd'].astype('category') # converting type of columns to 'category'
new_data['Fuel Metering Sys Cd'] = new_data['Fuel Metering Sys Cd'].cat.codes # Assigning numerical values and storing in another column

# Label encoding for Fuel Metering Sys Desc
new_data['Fuel Metering Sys Desc'] = new_data['Fuel Metering Sys Desc'].astype('category') # converting type of columns to 'category'
new_data['Fuel Metering Sys Desc'] = new_data['Fuel Metering Sys Desc'].cat.codes # Assigning numerical values and storing in another column

# Frequency/count encoding for Oil Viscosity
encoder = ce.CountEncoder(cols=['Oil Viscosity'])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Label encoding for Stop/Start System (Engine Management System) Code 
encoder = ce.OrdinalEncoder(mapping=[{'col': 'Stop/Start System (Engine Management System) Code', 
                                      'mapping': {'N': 0, 'Y': 1}}])
encoder.fit(new_data)
new_data = encoder.transform(new_data)

# Dealing with missing values (NaN)
new_data['Max Ethanol % - Gasoline'] = new_data['Max Ethanol % - Gasoline'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['Max Biodiesel %'] = new_data['Max Biodiesel %'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['2Dr Pass Vol'] = new_data['2Dr Pass Vol'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['2Dr Lugg Vol'] = new_data['2Dr Lugg Vol'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['4Dr Pass Vol'] = new_data['4Dr Pass Vol'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['4Dr Lugg Vol'] = new_data['4Dr Lugg Vol'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['Htchbk Pass Vol'] = new_data['Htchbk Pass Vol'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['Htchbk Lugg Vol'] = new_data['Htchbk Lugg Vol'].replace(np.nan, 0)  # replacing NaN values with 0
new_data['GHG 1-10 rating on Ethanol (EPA Determined)'] = new_data['GHG 1-10 rating on Ethanol (EPA Determined)'].replace(np.nan, 0)  # replacing NaN values with 0

In [665]:
# Feature extration(feature enginerring)

new_data['Transmission'] = new_data['Transmission'].str.lower() # Converting string to lower case
new_data['Auto_transmission'] = new_data['Transmission'].str.contains("auto") # Searching for the auto word
new_data=new_data.drop(labels='Transmission',axis=1) # Dropping original feature(Transmission)

In [666]:
new_data.transpose() # Looking at the data by interchanging rows and columns 

0        1        2     \
Mfr Name                                                 64       91        6   
Division                                                 18        7        6   
Carline                                                   1        1        1   
Verify Mfr Cd                                            64       91        6   
Index (Model Type Index)                                 41      403        3   
Eng Displ                                               3.5      1.8        4   
# Cyl                                                     6        4        8   
Comb Unadj FE - Conventional Fuel                   29.3561  34.4702  27.3986   
Air Aspiration Method Desc                                2        2        2   
Trans Desc                                                4        2        3   
# Gears                                                   9        6        8   
Lockup Torque Converter                                   0        0        0   
Trans Creeper Gear                                        1        1        1   
Drive Sys                                                 0        1        1   
Max Ethanol % - Gasoline                                 10       10       10   
Max Biodiesel %                                           0        0        0   
Fuel Usage Desc - Conventional Fuel                       1        2        2   
Gas Guzzler Exempt (Where Truck = 1975 NHTSA tr...        1        1        1   
2Dr Pass Vol                                              0        0        0   
2Dr Lugg Vol                                              0        0        0   
4Dr Pass Vol                                              0        0        0   
4Dr Lugg Vol                                              0        0        0   
Htchbk Pass Vol                                           0        0        0   
Htchbk Lugg Vol                                           0        0        0   
Intake Valves Per Cyl                                     2        2        2   
Exhaust Valves Per Cyl                                    2        2        2   
Carline Class                                             1        1        1   
Carline Class Desc                                       85       85       85   
Calc Approach Desc                                     1043      132       95   
Cyl Deact?                                                1        1        1   
Var Valve Timing?                                         1        1        1   
Var Valve Timing Desc                                    81       17       48   
Var Valve Lift?                                           0        0        0   
Fuel Metering Sys Desc                                    2        2        1   
Fuel Metering Sys Cd                                      1        1        3   
Oil Viscosity                                             9       15       47   
Stop/Start System (Engine Management System) Code         1        0        0   
FE Rating (1-10 rating on Label)                          4        6        4   
GHG Rating (1-10 rating on Label)                         4        6        4   
GHG 1-10 rating on Ethanol (EPA Determined)               0        0        0   
#1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Lab...        3        1        3   
City CO2 Rounded Adjusted                               419      365      494   
Hwy CO2 Rounded Adjusted                                406      259      364   
Comb CO2 Rounded Adjusted (as shown on FE Label)        420      317      435   
Auto_transmission                                      True     True     True   

                                                       3        4       5     \
Mfr Name                                                 69      122      69   
Division                                                 33       87       1   
Carline                                                   1        1       1   
Verify M

In [667]:
new_data.isnull().sum() # checking number of null values

Mfr Name                                                               0
Division                                                               0
Carline                                                                0
Verify Mfr Cd                                                          0
Index (Model Type Index)                                               0
Eng Displ                                                              0
# Cyl                                                                  0
Comb Unadj FE - Conventional Fuel                                      0
Air Aspiration Method Desc                                             0
Trans Desc                                                             0
# Gears                                                                0
Lockup Torque Converter                                                0
Trans Creeper Gear                                                     0
Drive Sys                                          

### <font color = green> Objective </font>
#### The main objective of this project is to predict 'Unadjusted combined fuel economy'(Comb Unadj FE - Conventional Fuel), which is a weighted average of city fuel economy and highway economy (55% and 45%, respectively).


### <font color = green> Deliverables: </font>
#### A clean dataset with all numeric and no null values, which can be used to train a Random forest model.

### <font color = green> Steps used for preparing the data </font>
- #### The columns we removed were either of the following - 
 ####  1) Did not carry any information useful for the model (example: Model year, index, or Fuel Unit Desc - Conventional Fuel)
 ####  2) They are two similar columns that were identical, so we removed one of them (example - Air Aspir Method and Air Aspiration             Method Desc)    
 ####  3)Not related to the objective or indirectly giving target information (ex - Hwy Unadj FE - Conventional Fuel)

- #### Depending upon the data type, variable type(ordinal or nominal), and the number of categories, we choose the kind of encoding to be used. We also took help from the lecture notes and a few websites when the answer was not that obvious. 


- #### If the feature had different categories with no order, we used label encoding or frequency encoding to convert them into the numeric format. 
- #### Using frequency encoding for the nominal variable is not the best option as the model will assume that a variable with 1 label is less valuable than a variable with 2 labels. But on the flip side, we are avoiding using one-hot encoding which can lead to dimensionality problems, so it's a fair trade-off.

- #### If the feature had different categories with an order or possible order, we have used ordinal encoding. The label with the most importance is given the highest number and the least important label is given the smallest number.
- #### There were few cases where genreal order is difficult so we have made few assumptions. For instance, Trans Desc here we have assumed this order 'Continuously Variable'> 'Selectable CV' > 'Automatic' > 'Automated Manual- Selectable' > 'Semi-Automatic' > 'Automated Manual' > 'Manual'.




- #### Few features could be used for our model but had too many NaN or missing values. For such, we used the replace() function and filled them with appropriate values(ex - Max Biodiesel % )

- #### We used the 'Transmission' column for feature extraction. Generally, vehicles with auto transmission have a better economy. So we tried to extract whether the vehicle has auto or manual transmission and used ordinal encoding to make a new column 'Auto_transmission'.

### <font color = green> What's next</font>

#### Out of all these features, many are going to be duplicates or irrelevant to the target feature. To figure out which ones are useful and which are not, we will have to train a Random forest model and look at the following things- 
 #### 1) Out of bag score for R^2
 #### 2) Relative feature importance 
 #### 3) Number of tree nodes, and
 #### 4) Median tree depth

### <font color = green> Result: </font>
#### The final dataset has 45 features which are all numeric, no missing values, and ready to be used for the Random forest model.